src/renderer/metal.rs

```diff
@@ -0,0 +1,208 @@
+    MTLPixelFormat, MTLTexture,
+};
+use objc2_quartz_core::{CAMetalDrawable, CAMetalLayer};
+use skia_safe::{
+    gpu::{
+        self,
+        mtl::{BackendContext, TextureInfo},
+        surfaces::wrap_backend_render_target,
+        DirectContext, SurfaceOrigin,
+    },
+    Canvas, ColorSpace, ColorType, Surface,
+};
+use winit::{event_loop::EventLoopProxy, window::Window};
+
+use crate::{
+    profiling::tracy_gpu_zone,
+    renderer::{SkiaRenderer, VSync},
+    window::{macos::get_ns_window, UserEvent},
+};
+
+struct MetalDrawableSurface {
+    pub _drawable: Retained<ProtocolObject<dyn CAMetalDrawable>>,
+    pub metal_drawable: Retained<ProtocolObject<dyn MTLDrawable>>,
+    pub surface: Surface,
+}
+
+impl MetalDrawableSurface {
+    fn new(
+        drawable: Retained<ProtocolObject<dyn CAMetalDrawable>>,
+        context: &mut DirectContext,
+    ) -> MetalDrawableSurface {
+        tracy_gpu_zone!("MetalDrawableSurface.new");
+
+        let texture = unsafe { drawable.texture() };
+        let texture_info = unsafe { TextureInfo::new(Retained::as_ptr(&texture).cast()) };
+        let backend_render_target = gpu::backend_render_targets::make_mtl(
+            (texture.width() as i32, texture.height() as i32),
+            &texture_info,
+        );
+
+        let metal_drawable =
+            unsafe { Retained::cast::<ProtocolObject<dyn MTLDrawable>>(drawable.clone()) };
+
+        let surface = wrap_backend_render_target(
+            context,
+            &backend_render_target,
+            SurfaceOrigin::TopLeft,
+            ColorType::BGRA8888,
+            ColorSpace::new_srgb(),
+            None,
+        )
+        .expect("Failed to create skia surface with metal drawable.");
+
+        MetalDrawableSurface {
+            _drawable: drawable,
+            metal_drawable,
+            surface,
+        }
+    }
+
+    fn metal_drawable(&self) -> &ProtocolObject<dyn MTLDrawable> {
+        &self.metal_drawable
+    }
+}
+
+pub struct MetalSkiaRenderer {
+    window: Window,
+    _device: Retained<ProtocolObject<dyn MTLDevice>>,
+    command_queue: Retained<ProtocolObject<dyn MTLCommandQueue>>,
+    metal_layer: Retained<CAMetalLayer>,
+    _backend: BackendContext,
+    context: DirectContext,
+    metal_drawable_surface: Option<MetalDrawableSurface>,
+}
+
+impl MetalSkiaRenderer {
+    pub fn new(window: Window, srgb: bool, vsync: bool) -> Self {
+        log::info!("Initialize MetalSkiaRenderer...");
+
+        let draw_size = window.inner_size();
+        let ns_window = get_ns_window(&window);
+
+        let device = unsafe {
+            Retained::retain(MTLCreateSystemDefaultDevice())
+                .expect("Failed to create Metal system default device.")
+        };
+        let metal_layer = unsafe {
+            let metal_layer = CAMetalLayer::new();
+            metal_layer.setDevice(Some(&device));
+            metal_layer.setPixelFormat(if srgb {
+                MTLPixelFormat::BGRA8Unorm_sRGB
```

Comment:
The awkward behavior I got here is that not setting it at all for sRGB solves the problem I had with non vivid colors so maybe it could be better to only disable sRGB if requested to and avoid the BGRA8Unorm_sRGB due to compilation errors.

At least on M3 silicon it doesn't identify as an available option (81)

```rust
if !srgb {
    metal_layer.setPixelFormat(MTLPixelFormat::BGRA8Unorm)
};
```

this would be enough

Reply:
> At least on M3 silicon it doesn't identify as an available option (81)

I'm not sure. Based on https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf, M3 silicon should support `MTLPixelFormat::BGRA8Unorm_sRGB`.

The reason why I write this is simply that I guess it based on the **name**. Actually I have no clear idea about it. This is very naive, indeed.

I'm totally unfamiliar about sRGB things. And I need to spend some time on investigating it. But it may take some time for me.

Reply:
what I am sure is that it is getting an error for `MTLPixelFormat::BGRA8Unorm_sRGB`
```bash
fatal error: "Invalid MTLPixelFormat: 81
```

seems that we should *not* change the pixel format, once we handle the sRGB on our skia [surface](https://github.com/neovide/neovide/pull/2461/files#diff-bd0d9f27119e48a1f8ca5898037d7f50ee5325f8418e884a749514b165919ce3R48-R55)

another thing is that at least for me the `BGRA8Unorm` is the default pixel format and it sounds correct, setting the pixel format again here, make it loose the skia surface we have applied before somehow.

Reply:
maybe we should even remove the `metal_layer.setPixelFormat`

Reply:
Adding the sRGB support under the hood would be better as default on macOS 

```rust
unsafe {
    ns_window.setColorSpace(Some(NSColorSpace::deviceRGBColorSpace().as_ref()));
}
```

And I just tested again, and not setting the pixel format at all works perfectly fine here.

Reply:
> maybe we should even remove the metal_layer.setPixelFormat

I think we still need to explicitly specify the pixel format and align it with the one of surface, which is specified at about line 52, rather than depend on the default behavior, which might be changed by Apple in the future. Though change of default is very unlike to happen, we explicitly specify it does no harm.

```
let surface = wrap_backend_render_target(
    context,
    &backend_render_target,
    SurfaceOrigin::TopLeft,
    ColorType::BGRA8888,
    ColorSpace::new_srgb(),
    None,
)
```

> Adding the sRGB support under the hood would be better as default on macOS

I think this change is good and I'll apply it later. But I still have no clear idea about this because of a lack of knowledge on Apple's API.

Reply:
> Adding the sRGB support under the hood would be better as default on macOS
> 
> ```rust
> unsafe {
>     ns_window.setColorSpace(Some(NSColorSpace::deviceRGBColorSpace().as_ref()));
> }
> ```
> 
> And I just tested again, and not setting the pixel format at all works perfectly fine here.

And according to https://developer.apple.com/documentation/quartzcore/cametallayer/1478170-colorspace, maybe it is better to set `colorspace` on `CAMetalLayer` instead of on `NSWindow`? Just my guess.

Reply:
As an example Alacrity is using `NSWindow` to specify it. https://github.com/alacritty/alacritty/pull/6602

Reply:
> I think we still need to explicitly specify the pixel format and align it with the one of surface, which is specified at about line 52, rather than depend on the default behavior, which might be changed by Apple in the future. Though change of default is very unlike to happen, we explicitly specify it does no harm.

specifying the `metal_layer.setPixelFormat` as `BGRA8Unorm_sRGB` get's an error as I mentioned. IMO applying `BGRA8Unorm_sRGB` get's no difference but try to fix that, otherwise just apply the `BGRA8Unorm` to disable the sRGB.

Reply:
@falcucci 

> As an example Alacrity is using NSWindow to specify it. https://github.com/alacritty/alacritty/pull/6602

That's good. I'll apply it. And I will also read some codes of alacritty to make this behavior correct. It is a very nice way to handle this by following alacritty.

> specifying the metal_layer.setPixelFormat as BGRA8Unorm_sRGB get's an error as I mentioned. IMO applying BGRA8Unorm_sRGB get's no difference but try to fix that, otherwise just apply the BGRA8Unorm to disable the sRGB.

Sorry, I might mislead you. I mean set `BGRA8Unorm` always explicitly rather than rely on default behavior. Though very unlikely, the default value might change by Apple in the future. And this format must align with skia surface's one. Explicitly doing this makes it clearer.

Reply:
setting `BGRA8Unorm` explicitly breaks the sRGB surface, and the colors are no vivid anymore.

Reply:
I'm don't know how it currently looks like, but for this PR, I think we should try to keep the colors exactly the same as before. 

We can make the `srgb` and `no-srgb` option do as expected in another PR. Probably so that `no-srgb` is the default, and emulates the old Neovide behaviour, while `srgb` has more dull colors, and matches the `iterm2` colors.

Reply:
@fredizzimo Agree. that's why I am suggesting this

```rust
if !srgb {
    metal_layer.setPixelFormat(MTLPixelFormat::BGRA8Unorm)
};
```

to keep the default behavior.

we can troubleshoot anything else afterwards.

Reply:
> I'm don't know how it currently looks like, but for this PR, I think we should try to keep the colors exactly the same as before.

@fredizzimo Agree with you, actually. This pr should never be merged until we get clear on this problem totally. At least make it work.

Though I never use color space personally, I think many people care about it very much.

Reply:
Done! Please have a double check on functions both for --no-srgb and --srgb. Thanks!